examples/07_end_to_end/seismic_waveform.ipynb

```diff
@@ -0,0 +1,1325 @@
+    "    basket = fdsn_waveforms(client, \"TA\", \"E42A\", \"*\", \"BH?\", timedelta(seconds=4))\n",
+    "    for k, v in basket.items():\n",
+    "        csp.print(k, v)\n",
+    "\n",
+    "out = csp.run(graph, starttime=starttime, endtime=endtime)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "85bfc73c-db25-44fa-a0a5-02ac22ca0b5f",
+   "metadata": {},
+   "source": [
+    "## Real-time Data via Seedlink\n",
+    "In this section, we build a real time adapter to live waveform data. It is straightforward to apply the filtering and triggering logic from the previous section to these outputs."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "96fc1fee-0d20-45f5-8724-b6905dcbcd54",
+   "metadata": {},
+   "source": [
+    "### Real-time Adapter Definition"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "id": "4808fc0e-53a1-4677-aa28-8542ce8f31f5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import logging\n",
+    "import threading\n",
+    "from fnmatch import fnmatch\n",
+    "from obspy.clients.seedlink.client.seedlinkconnection import SeedLinkConnection\n",
+    "from obspy.clients.seedlink.client.slstate import SLState\n",
+    "from obspy.clients.seedlink.slpacket import SLPacket\n",
+    "from urllib.parse import urlparse\n",
+    "\n",
+    "from csp.impl.adaptermanager import AdapterManagerImpl\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "\n",
+    "class EasySeedLinkAdapter(AdapterManagerImpl):\n",
+    "    \"\"\"Real time push adapter (technically, an adapter manager) that manages subscription to multiple seedlink streams for the same connection.\n",
+    "\n",
+    "    See examples/04_writing_adapters/e7_adaptermanager_inputoutput.py for the implementation this was based on. \n",
+    "    \"\"\"\n",
+    "    \n",
+    "    def __init__(self, server_url: str):\n",
+    "        \"\"\" Constructor\n",
+    "        \n",
+    "        Args:\n",
+    "            server_url: The url and port of the seedlink server we connect to.\n",
+    "        \"\"\"\n",
+    "        self._server_url = server_url\n",
+    "        parsed_url = urlparse(server_url, scheme=\"seedlink\")\n",
+    "        self._conn = SeedLinkConnection()\n",
+    "        self._conn.set_sl_address(\"%s:%d\" % (parsed_url.hostname, parsed_url.port or 18000))\n",
+    "\n",
+    "        # We will keep track of requested input adapters here\n",
+    "        self._inputs = {}\n",
+    "\n",
+    "        # Out driving thread, all  realtime adapters will need a separate thread of execution that drives data into the engine thread\n",
+    "        self._running = False\n",
+    "        self._thread = None\n",
+    "        self._starttime = None\n",
+    "        self._endtime = None\n",
+    "\n",
+    "    def subscribe(self, network, station, selector=None, push_mode=csp.PushMode.NON_COLLAPSING):\n",
```

Comment:
this isnt the correct way to expose adapter managers.  This is the impl ( runtime ) instance that should only be created by the engine when it starts running.
It shouldnt have the user facing subscribe API, which should be a separate light weight "graph time" object

Reply:
I changed this back,